doc/design/thin_pool.md

```diff
@@ -0,0 +1,123 @@
+# LVMO: Thin provisioning
+
+## Summary
+- LVM thin provisioning allows creation of volumes whose combined size is greater than that of the available storage.
+
+**Advantages**:
+- Storage space can be used more effectively. More users can be accommodated for the same amount of storage space when compared to thick provisioning. This significantly reduces upfront hardware cost for the storage admins.
+- Faster clones and snapshots
+
+**Disadvantages** :
+- Reduced performance when compared to thick volumes.
+- Over-allocation of the space. This can be mitigated by better monitoring of the disk usage.
+
+LVM does this by allocating the blocks in a thin LV from a special "thin pool LV". A thin pool LV must be created before thin LVs can be created within it.
+
+The LVMO will create a thin-pool LV in the volume group in order to create thinly provisioned volumes.
+
+
+## Proposal:
+- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a volume group and a thin-pool.
+- One volume group will be mapped to a single thin pool.
+- User should be able to configure the thin-pool size based on percentage of the available volume group size.
+- Default chunk size of the thin pool will be 512 kiB
+- `lvmd.yaml` config file should be updated with the device class, volume group and thin-pool mapping.
+- Alerts should be triggered if a thin-pool `data` or `metadata` available size crosses a predefined threshold limit.
+
+
+## Design Details
+### API changes:
+
+- `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig` will have the mapping between device class, volume group and the thin-pool.
+- One DeviceClass can be mapped to only one thin-pool.
+
+- `LVMCluster` API changes
+```go=
++ type ThinPoolConfig struct{
++    // Name of the thin pool to be creatd
++	Name string `json:"name,omitempty"`
+
++   // SizePercent represents percentage of remaining space in the volume group that should be used
++	// for creating the thin pool.
++	// +kubebuilder:validation:default=75
++	// +kubebuilder:validation:Minimum=10
+	// +kubebuilder:validation:Maximum=90
++	SizePercent int `json:"sizePercent,omitempty"`
+
++	// OverProvisionRatio represents the ratio of overprovision that can
++	// be allowed on thin pools
++	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
+}
+
+type DeviceClass struct {
+    Name string `json:"name,omitempty"`
+
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
```

Comment:
- maybe use a pointer which helps in checking against `nil` for absence and drop the omitemtpy as well.